object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java

```diff
@@ -0,0 +1,515 @@
+
+  /** By default, should we transfer ownership to the caller when a variable is returned? */
+  static final boolean TRANSFER_OWNERSHIP_AT_RETURN = true;
+
+  /** {@code @MustCall} errors reported thus far, to avoid duplicates */
+  private final Set<LocalVarWithTree> reportedMustCallErrors = new HashSet<>();
+
+  private final ObjectConstructionAnnotatedTypeFactory typeFactory;
+
+  private final BaseTypeChecker checker;
+
+  private final CFAnalysis analysis;
+
+  MustCallInvokedChecker(
+      ObjectConstructionAnnotatedTypeFactory typeFactory,
+      BaseTypeChecker checker,
+      CFAnalysis analysis) {
+    this.typeFactory = typeFactory;
+    this.checker = checker;
+    this.analysis = analysis;
+  }
+
+  /**
+   * This function traverses the given method CFG and reports an error if "f" isn't called on any
+   * local variable node whose class type has @MustCall(f) annotation before the variable goes out
+   * of scope. The traverse is a standard worklist algorithm. Worklist and visited entries are
+   * BlockWithLocals objects that contain a set of (LocalVariableNode, Tree) pairs for each block. A
+   * pair (n, T) represents a local variable node "n" and the latest AssignmentTree "T" that assigns
+   * a value to "n".
+   *
+   * @param cfg the control flow graph of a method
+   */
+  void checkMustCallInvoked(ControlFlowGraph cfg) {
+    // add any owning parameters to initial set of variables to track
+    BlockWithLocals firstBlockLocals =
+        new BlockWithLocals(cfg.getEntryBlock(), computeOwningParameters(cfg));
+
+    Set<BlockWithLocals> visited = new HashSet<>();
+    Deque<BlockWithLocals> worklist = new ArrayDeque<>();
+
+    worklist.add(firstBlockLocals);
+    visited.add(firstBlockLocals);
+
+    while (!worklist.isEmpty()) {
+
+      BlockWithLocals curBlockLocals = worklist.removeLast();
+      List<Node> nodes = curBlockLocals.block.getNodes();
+      // defs to be tracked in successor blocks, updated by code below
+      Set<LocalVarWithTree> newDefs = new HashSet<>(curBlockLocals.localSetInfo);
+
+      for (Node node : nodes) {
+
+        if (node instanceof AssignmentNode) {
+          handleAssignment((AssignmentNode) node, newDefs);
+        } else if (node instanceof ReturnNode) {
+          handleReturn((ReturnNode) node, cfg, newDefs);
+        } else if (node instanceof MethodInvocationNode || node instanceof ObjectCreationNode) {
+          handleInvocation(newDefs, node);
+        }
+      }
+
+      handleSuccessorBlocks(visited, worklist, newDefs, curBlockLocals.block);
+    }
+  }
+
+  private void handleInvocation(Set<LocalVarWithTree> newDefs, Node node) {
+    List<Node> arguments;
+    ExecutableElement executableElement;
+    if (node instanceof MethodInvocationNode) {
+      MethodInvocationNode invocationNode = (MethodInvocationNode) node;
+      arguments = invocationNode.getArguments();
+      executableElement = TreeUtils.elementFromUse(invocationNode.getTree());
+    } else {
```

Comment:
I'd add a test here that this is actually an `ObjectConstructionNode`, and throw a `BugInCF` in the actual `else` block, just to be defensive.

Reply:
e48f172